File: app/src/main/java/com/scop/org/minesweeper/generators/solver/DeepSolver.java
package com.scop.org.minesweeper.generators.solver;

import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static com.scop.org.minesweeper.elements.Tile.Status.*;

public class DeepSolver extends Solver {
	private static final int NONE = -1;
	private static final int CASE_1_1 = 1;
	private static final int CASE_2_1 = 2;

	@Override
	public boolean analyze(){
		boolean changesMade = false;

		List<Tile> sketchTiles = sketch.getTiles();

		for (Integer i : sketch.getNumberedCopy()) {
			Tile sTile = sketchTiles.get(i);

			if (sTile.isNumberVisible()) {
				List<Integer> coveredIdx = GridUtils.getNeighborsIdx(sketch.getGrid(), sTile).stream()
					.filter(idx->sketchTiles.get(idx).getStatus() == COVERED)
					.collect(Collectors.toList());

				if (coveredIdx.size() > 0) {
					int bNear = sTile.getBombsNear();
					int totalCovers = coveredIdx.size();
					coveredIdx.stream()
							.map(sketchTiles::get)
							.forEach(t->t.setCustomFlag('?'));

					List<Tile> numberedNeighbors =  GridUtils.getNeighbors(sketch.getGrid(),sTile).stream()
							.filter(Tile::isNumberVisible)
							.collect(Collectors.toList());

					for (Tile sTile2 : numberedNeighbors) {
						List<Integer> coveredIdx2 = GridUtils.getNeighborsIdx(sketch.getGrid(), sTile2).stream()
								.filter(idx->sketchTiles.get(idx).getStatus() == COVERED)
								.collect(Collectors.toList());

						int bNear2 = sTile2.getBombsNear();
						int totalCovers2 = coveredIdx2.size();
						int count = (int) coveredIdx2.stream()
								.filter(idx->sketchTiles.get(idx).hasCustomFlag('?'))
								.count();
						if (count > 0) {
							int mCase = matchLogic(bNear, bNear2, totalCovers, totalCovers2, count);
							switch (mCase) {
								case CASE_2_1:
									for (int idx : coveredIdx){
										if (!coveredIdx2.contains(idx)){
											this.markBomb(idx);
										}
									}

								case CASE_1_1:
									for (int idx : coveredIdx2){
										if (!coveredIdx.contains(idx)){
											this.reveal(idx);
										}
									}

									changesMade = true;
									break;
							}
						}
					}
					coveredIdx.stream()
							.map(sketchTiles::get)
							.forEach(t->t.removeCustomFlag('?'));
				}
			}
		}

		return changesMade;
	}

	private int matchLogic(int bNear, int bNear2, int totalCovers, int totalCovers2, int count) {
		if (bNear > bNear2) {
			int diff = bNear-bNear2;
			return totalCovers-diff == count? CASE_2_1:NONE;
		}
		if (bNear == bNear2){
			return count == totalCovers? (totalCovers == totalCovers2? NONE:CASE_1_1):NONE;
		}
		return NONE;
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/solver/DeepSolver.kt
package org.oar.minesweeper.generators.solver

import org.oar.minesweeper.models.Tile
import org.oar.minesweeper.models.TileStatus
import org.oar.minesweeper.utils.GridUtils.getNeighbors
import org.oar.minesweeper.utils.GridUtils.getNeighborsIdx

class DeepSolver : Solver() {

    companion object {
        private const val NONE = -1
        private const val CASE_1_1 = 1
        private const val CASE_2_1 = 2
    }

    override fun analyze(): Boolean {
        var changesMade = false
        val sketchTiles = sketch.tiles
        for (i in sketch.numberedCopy) {
            val sTile = sketchTiles[i]
            if (sTile.isNumberVisible) {
                val coveredIdx = sketch.grid.getNeighborsIdx(sTile)
                    .filter { idx -> sketchTiles[idx].status === TileStatus.COVERED }

                if (coveredIdx.isNotEmpty()) {
                    val bNear = sTile.bombsNear
                    val totalCovers = coveredIdx.size
                    coveredIdx
                        .map { idx -> sketchTiles[idx] }
                        .forEach { tile -> tile.customFlag.add('?') }

                    val numberedNeighbors = sketch.grid.getNeighbors(sTile)
                        .filter(Tile::isNumberVisible)

                    for (sTile2 in numberedNeighbors) {
                        val coveredIdx2 = sketch.grid.getNeighborsIdx(sTile2)
                            .filter { idx -> sketchTiles[idx].status === TileStatus.COVERED }

                        val bNear2 = sTile2.bombsNear
                        val totalCovers2 = coveredIdx2.size
                        val count = coveredIdx2
                            .count { idx: Int -> sketchTiles[idx].customFlag.contains('?') }

                        if (count > 0) {
                            when (matchLogic(bNear, bNear2, totalCovers, totalCovers2, count)) {
                                CASE_2_1 -> {
                                    for (idx in coveredIdx) {
                                        if (!coveredIdx2.contains(idx)) {
                                            markBomb(idx)
                                        }
                                    }
                                    for (idx in coveredIdx2) {
                                        if (!coveredIdx.contains(idx)) {
                                            reveal(idx)
                                        }
                                    }
                                    changesMade = true
                                }
                                CASE_1_1 -> {
                                    for (idx in coveredIdx2) {
                                        if (!coveredIdx.contains(idx)) {
                                            reveal(idx)
                                        }
                                    }
                                    changesMade = true
                                }
                            }
                        }
                    }
                    coveredIdx
                        .map { idx -> sketchTiles[idx] }
                        .forEach { tile -> tile.customFlag.remove('?') }
                }
            }
        }
        return changesMade
    }

    private fun matchLogic(
        bNear: Int,
        bNear2: Int,
        totalCovers: Int,
        totalCovers2: Int,
        count: Int
    ): Int {
        if (bNear > bNear2) {
            val diff = bNear - bNear2
            return if (totalCovers - diff == count) CASE_2_1 else NONE
        }
        return if (bNear == bNear2) {
            if (count == totalCovers) if (totalCovers == totalCovers2) NONE else CASE_1_1 else NONE
        } else NONE
    }
}